File: 543.java
class Solution {
    int result = 0;

    public int height(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int lh = height(root.left);
        int rh = height(root.right);
        result = Math.max(result, lh + rh);
        return 1 + Math.max(lh, rh);
    }

    public int diameterOfBinaryTree(TreeNode root) {
        height(root);
        return result;
    }
}

File: Python/543.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        maxD = 0

        def helper(root):
            nonlocal maxD
            if not root:
                return 0

            left = helper(root.left)
            right = helper(root.right)

            maxD = max(maxD, left + right)
            return 1 + max(left, right)

        helper(root)
        return maxD
